// @terra-dev/web-connector-interface/_commonjs/utils/serializeTx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeTx = exports.serializeTx = void 0;
const terra_js_1 = require("@terra-money/terra.js");
function serializeTx(tx) {
    var _a, _b, _c;
    return {
        msgs: tx.msgs.map((msg) => msg.toJSON()),
        fee: (_a = tx.fee) === null || _a === void 0 ? void 0 : _a.toJSON(),
        memo: tx.memo,
        gasPrices: (_b = tx.gasPrices) === null || _b === void 0 ? void 0 : _b.toString(),
        gasAdjustment: (_c = tx.gasAdjustment) === null || _c === void 0 ? void 0 : _c.toString(),
        account_number: tx.account_number,
        sequence: tx.sequence,
        feeDenoms: tx.feeDenoms,
    };
}
exports.serializeTx = serializeTx;
function deserializeTx(tx) {
    return {
        ...tx,
        msgs: tx.msgs.map((msg) => terra_js_1.Msg.fromData(JSON.parse(msg))),
        fee: tx.fee ? terra_js_1.StdFee.fromData(JSON.parse(tx.fee)) : undefined,
    };
}
exports.deserializeTx = deserializeTx;
