// @terra-money/wallet-provider/react/StaticWalletProvider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { WalletContext, WalletStatus, } from '@terra-dev/use-wallet';
import { useMemo } from 'react';
export function StaticWalletProvider({ children, defaultNetwork, status = WalletStatus.INITIALIZING, availableConnectTypes = [], availableInstallTypes = [], wallets = [], }) {
    const state = useMemo(() => {
        return {
            availableConnectTypes,
            availableInstallTypes,
            status,
            network: defaultNetwork,
            wallets,
            install: () => {
                throw new Error('not implemented!');
            },
            connect: () => {
                throw new Error('not implemented!');
            },
            connectReadonly: () => {
                throw new Error('not implemented!');
            },
            disconnect: () => {
                throw new Error('not implemented!');
            },
            post: () => {
                throw new Error('not implemented!');
            },
            sign: () => {
                throw new Error('not implemented!');
            },
            recheckStatus: () => {
                throw new Error('not implemented!');
            },
            isChromeExtensionCompatibleBrowser: () => {
                throw new Error('not implemented!');
            },
        };
    }, [
        availableConnectTypes,
        availableInstallTypes,
        defaultNetwork,
        status,
        wallets,
    ]);
    return (_jsx(WalletContext.Provider, Object.assign({ value: state }, { children: children }), void 0));
}
